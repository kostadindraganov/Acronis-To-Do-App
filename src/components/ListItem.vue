<template>
  <el-card
    v-loading="loading"
    shadow="hover"
    class="card-item-wrapper"
    @mouseenter.native="mouseOver = true"
    @mouseleave.native="mouseOver = false"
    :class="{'bg-on': mouseOver, 'bg-off': !mouseOver}"
  >
    <el-row :gutter="0" type="flex" justify="space-between" align="middle">
      <el-col :xs="1" :sm="1" :md="1" :lg="1" :xl="1" class="item-left">
        <el-checkbox @change="editItem" :checked="isDone" v-model="checked" size="medium"></el-checkbox>
      </el-col>
      <el-col :xs="19" :sm="22" :md="22" :lg="22" :xl="22" class="item-left">
        <span>{{itemTitle}}</span>
      </el-col>
      <el-col :xs="1" :sm="1" :md="1" :lg="1" :xl="1" class="item-right">
        <transition name="el-fade-in-linear">
          <i v-show="mouseOver" class="el-icon-close" @click="deleteItem"></i>
        </transition>
      </el-col>
    </el-row>
  </el-card>
</template>

<script>
  import {
    GET_DELETE_ITEM,
    GET_CURRENT_LIST,
    GET_EDIT_ITEM
  } from "@/store/mutation-types";
  export default {
    name: "ListItem",
    props: {
      itemId: {
        type: Number,
        required: true,
        default: 1,
        useDefaultForNull: true
      },
      itemTitle: {
        type: String,
        required: true,
        default: "",
        useDefaultForNull: true
      },
      isDone: {
        type: Boolean,
        required: true,
        default: false,
        useDefaultForNull: true
      }
    },
    data: () => ({
      checked: false,
      loading: false,
      mouseOver: false
    }),
    methods: {
      deleteItem() {
        this.loading = true;
        this.$store.dispatch(GET_DELETE_ITEM, this.itemId).then(res => {
          if (res) {
            ///check response return status ok
          }
          this.loading = false;
          this.$store.dispatch(GET_CURRENT_LIST);
        });
      },
      editItem() {
        this.loading = true;
        const item = { id: this.itemId, title: this.title, state: this.checked };
        this.$store.dispatch(GET_EDIT_ITEM, item).then(res => {
          if (res) {
            ///check response return status ok
          }
          this.loading = false;
          this.$store.dispatch(GET_CURRENT_LIST);
        });
      }
    }
  };
</script>
<style  lang="scss" scoped>
  .item-left {
    color: #727272;
    line-height: 20px;
    text-align: left;
    padding-left: 0px;
  }
  .item-right {
    text-align: right;
  }
  .el-icon-close {
    cursor: pointer;
  }
  .bg-on {
    background-color: #3b8de240;
  }
</style>


