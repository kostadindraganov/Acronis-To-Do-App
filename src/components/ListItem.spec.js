import { config } from "@vue/test-utils";
import { createLocalVue, shallowMount, mount } from "@vue/test-utils";
import Vuex from "vuex";
import ElementUI from "element-ui";
import ListItem from "@/components/ListItem";

config.stubs.transition = false;
const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(ElementUI);

describe("ListItem.vue", () => {
	let actions;
	let store;
	let wrapper;
	beforeEach(() => {
		actions = {
			getCurrentList: jest.fn(),
			getAddItem: jest.fn(),
			getEditItem: jest.fn(),
			getDeleteItem: jest.fn()
		};

		store = new Vuex.Store({
			state: {
				currentList: []
			},
			actions
		});

		wrapper = mount(ListItem, {
			localVue,
			store,
			propsData: {
				itemId: 1,
				itemTitle: "test",
				isDone: false
			}
		});
	});

	it("check props", () => {
		expect(wrapper.vm.itemId).toBe(1);
		expect(wrapper.vm.itemTitle).toBe("test");
		expect(wrapper.vm.isDone).toBe(false);
		wrapper.setProps({ itemId: "1" });
		wrapper.setProps({ itemTitle: 1 });
		wrapper.setProps({ isDone: "" });
		expect(Object.is(wrapper.vm.itemId, 1)).toBe(false);
		expect(Object.is(wrapper.vm.itemTitle, "1")).toBe(false);
		expect(Object.is(wrapper.vm.isDone, false)).toBe(false);
	});

	it("if checkbox dispatch an action", () => {
		wrapper.find(".el-checkbox").trigger("click");
		expect(actions.getEditItem).toHaveBeenCalled();
	});

	it("if delete item dispatch an action", () => {
		wrapper.find(".el-icon-close").trigger("click");
		expect(actions.getDeleteItem).toHaveBeenCalled();
	});

	test("renders correctly", () => {
		expect(wrapper.element).toMatchSnapshot();
	});
});
