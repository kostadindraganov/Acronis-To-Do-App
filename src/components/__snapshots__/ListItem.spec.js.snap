// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem.vue renders correctly 1`] = `
<div
  class="el-card card-item-wrapper is-hover-shadow bg-off"
>
  <!---->
  <div
    class="el-card__body"
  >
    <div
      class="el-row is-justify-space-between is-align-middle el-row--flex"
    >
      <div
        class="item-left el-col el-col-24 el-col-xs-1 el-col-sm-1 el-col-md-1 el-col-lg-1 el-col-xl-1"
      >
        <label
          class="el-checkbox"
          role="checkbox"
        >
          <span
            aria-checked="mixed"
            class="el-checkbox__input"
          >
            <span
              class="el-checkbox__inner"
            />
            <input
              aria-hidden="true"
              class="el-checkbox__original"
              type="checkbox"
              value=""
            />
          </span>
          <!---->
        </label>
      </div>
       
      <div
        class="item-left el-col el-col-24 el-col-xs-19 el-col-sm-22 el-col-md-22 el-col-lg-22 el-col-xl-22"
      >
        <span>
          test
        </span>
      </div>
       
      <div
        class="item-right el-col el-col-24 el-col-xs-1 el-col-sm-1 el-col-md-1 el-col-lg-1 el-col-xl-1"
      >
        <i
          class="el-icon-close"
          style="display: none;"
        />
      </div>
    </div>
  </div>
</div>
`;
